### PR TITLE
Feature/un 101 Translate blocktags for proxy calls

### PR DESCRIPTION
**Pull request by kudlaj**
- If a query is triggered with “latest” block, the “unode” will process it in the latest block it has processed, not the latest block on the ownershipchain.

- If a query is triggered with a block number larger than the latest processes block by the uNode it fails in the same way that a regular node will fail if you ask for a block not processed yet.

### CODE REVIEW

File: cmd/server/api/rpcmethods.go [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+}
+
+var rpcMethodsWithHash = map[string]RPCMethod{
+	"eth_getBlockByHash":                    RPCMethodEthGetBlockByHash,
+	"eth_getTransactionReceipt":             RPCMethodEthGetTransactionReceipt,
+	"eth_getTransactionByHash":              RPCMethodEthGetTransactionByHash,
+	"eth_getTransactionByBlockHashAndIndex": RPCMethodEthGetTransactionByBlockHashAndIndex,
+}
+
+//	var rpcMethodsWithCountByHash = map[string]RPCMethod{
```
Comment (zorjak) [inline cmd/server/api/rpcmethods.go:119]: should we remove this comment if it is not needed? or at least write a comment on why this is commented (not supported)

**Review by zorjak — APPROVED** [review]

File: cmd/server/api/rpcmethods.go [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+	ReplaceBlockTag(req *JSONRPCRequest, method RPCMethod, blockNumberUnode string) (*JSONRPCRequest, error)
+}
+
+type ProxyRPCMethodManager struct{}
+
+func NewProxyRPCMethodManager() RPCMethodManager {
+	return &ProxyRPCMethodManager{}
+}
+
+func (b *ProxyRPCMethodManager) HasRPCMethodWithBlocknumber(methodName string) (RPCMethod, bool) {
```
Comment (AlexRipoll) [inline cmd/server/api/rpcmethods.go]: Camel case typo in `HasRPCMethodWithBlocknumber`

**Review by AlexRipoll — CHANGES_REQUESTED** [review]

File: cmd/server/api/rpcmiddleware.go [hunk 3]
```diff
@@ -15,7 +15,7 @@ type JSONRPCRequest struct {
 	ID      *json.RawMessage  `json:"id,omitempty"`
 }
 
-type ParamsRPCRequest struct {
+type EthCallParamsRPCRequest struct {
```
Comment (luispdm) [inline cmd/server/api/rpcmiddleware.go]: If you do not plan to use it in external packages tests, this should be unexported

File: cmd/server/api/rpcmethods.go [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+)
+
+type RPCMethod int
+
+type BlockTag string
```
Comment (luispdm) [inline cmd/server/api/rpcmethods.go]: Make it unexported, as well as: FilterObject, Block, Transaction and the constants Latest, Pending etc. and CompareHex(), please

File: cmd/server/api/rpcproxyhandler.go [hunk 5]
```diff
@@ -49,6 +64,19 @@ func (h *RPCProxyHandler) HandleProxyRPC(r *http.Request, req JSONRPCRequest) RP
 	if err != nil {
 		return getErrorResponse(fmt.Errorf("error getting JSON RPC response: %w", err), req.ID)
 	}
+	// check if we have to replace the block tag
```
Comment (luispdm) [inline cmd/server/api/rpcproxyhandler.go]: I think this comment refers to linees 16-26 instead of 68-78

File: cmd/server/api/rpcmethods.go [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+	case *Block:
+		return value.Number, nil
+	case *Transaction:
+		return value.BlockNumber, nil
+	}
+
+	return "", fmt.Errorf("unknown type for unmarshalling")
+}
+
+func replaceBlockTagWithHash(req *JSONRPCRequest, position int, blockNumberHash string) error {
```
Comment (luispdm) [inline cmd/server/api/rpcmethods.go]: Rename "blockNumberHash" to "blockNumberUnode" here, as well as in "replaceBlockTagFromObject" and "getBlockNumber", please.

I also think that this function should be called: "replaceBlockTagWithBlockNumber".

**Review by luispdm — CHANGES_REQUESTED** [review]

**Review by AlexRipoll — APPROVED** [review]

**Review by luispdm — APPROVED** [review]

**Pull request merged**